spec/kinesis/enhanced_shard_reader_spec.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+require 'rspec'
+require 'spec_helper'
+require 'kinesis/enhanced_shard_reader'
+
+describe Kinesis::EnhancedShardReader do
+  let(:error_queue) { Queue.new }
+  let(:record_queue) { Queue.new }
+  let(:logger) { double('Logger', info: nil, error: nil) }
+  let(:kinesis_client) { instance_double(Aws::Kinesis::Client) }
+  let(:subscription) { double('Subscription', close: nil) }
+  let(:event_stream) { double('EventStream') }
+
+  before do
+    # Mock the SubthreadLoop behavior
+    allow_any_instance_of(Kinesis::SubthreadLoop).to receive(:run) do |instance|
+      # Call process once to simulate the loop
+      instance.send(:process)
+    end
+
+    # Mock the subscription behavior
+    allow(kinesis_client).to receive(:subscribe_to_shard).and_return(subscription)
+    allow(subscription).to receive(:on_event_stream).and_yield(event_stream)
+    allow(event_stream).to receive(:on_record_event)
+    allow(event_stream).to receive(:on_error_event)
+    allow(subscription).to receive(:wait)
+  end
+
+  subject do
+    described_class.new(
+      error_queue: error_queue,
+      logger: logger,
+      record_queue: record_queue,
+      shard_id: 'test-shard-id',
+      sleep_time: 0.1,
+      kinesis_client: kinesis_client,
+      consumer_arn: 'test-consumer-arn',
+      starting_position: { type: 'LATEST' }
+    )
+  end
+
+  describe '#initialize' do
+    it 'sets up a subscription to the shard' do
+      expect(kinesis_client).to receive(:subscribe_to_shard).with(
+        consumer_arn: 'test-consumer-arn',
+        shard_id: 'test-shard-id',
+        starting_position: { type: 'LATEST' }
+      )
+
+      # Trigger the process method
+      subject.send(:process)
+    end
+  end
+
+  describe '#shutdown' do
+    it 'closes the subscription' do
+      # Create the reader and set the subscription
+      reader = subject
+      reader.instance_variable_set(:@subscription, subscription)
+
+      # Expect the subscription to be closed
+      expect(subscription).to receive(:close)
+
+      reader.shutdown
+    end
+  end
+
+  describe '#process_records' do
+    it 'adds records to the queue' do
+      record = double('Record')
```

Comment:
nit: is this an actual class? would prefer `instance_double` if so

Reply:
done! I have also replaced the other `double` usages with `instance_double`